sound/soc/sof/intel/hda-stream.c

```diff
@@ -85,6 +85,60 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
 	return offset;
 }
 
+/*
+ * set up Buffer Descriptor List (BDL) for DMA trace host memory transfer
+ * BDL describes the location of the individual buffers and is little endian.
+ */
+static int hda_dsp_trace_setup_bdl(struct snd_sof_dev *sdev,
+				   struct snd_dma_buffer *dmab,
+				   struct hdac_stream *stream)
+{
+	struct sof_intel_dsp_bdl *bdl;
+	int offset, size;
+
+	/* program the initial BDL entries */
+	bdl = (struct sof_intel_dsp_bdl *)stream->bdl.area;
+	offset = 0;
+	stream->frags = 0;
+
+	if (!sdev->first_boot && sdev->trace_init_offset) {
+		/* set up BDL for [trace_init_offset, buffer_size) */
+		size = stream->bufsize - sdev->trace_init_offset;
+		offset = hda_setup_bdle(sdev, dmab,
+					stream, &bdl,
+					sdev->trace_init_offset,
+					size, 0);
+		if (offset != stream->bufsize) {
+			dev_err(sdev->dev, "error: bdl[0] setting for dma trace wrong!\n");
+			return -EINVAL;
+		}
+
+		/* set up BDL for [0, trace_init_offset) */
+		offset = hda_setup_bdle(sdev, dmab,
+					stream, &bdl, 0,
+					sdev->trace_init_offset, 0);
+		if (offset != sdev->trace_init_offset) {
+			dev_err(sdev->dev, "error: bdl[1] setting for dma trace wrong!\n");
+			return -EINVAL;
+		}
+
+	} else {
+		/*
+		 * for first boot or trace_init_offset=0 cases, we need
+		 * assign only one BDL entry for DMA trace
+		 */
+		offset = hda_setup_bdle(sdev, dmab,
+					stream, &bdl, 0,
+					stream->bufsize, 0);
+		if (offset != stream->bufsize) {
+			dev_err(sdev->dev, "error: bdl setting for dma trace wrong!\n");
```

Comment:
Btw, how is the value "wrong". Can you make the error messages more helpful for debug.

Reply:
OK, let me improve this.